haystack/pipelines/base.py

```diff
@@ -373,12 +390,17 @@ def add_node(self, component: BaseComponent, name: str, inputs: List[str]):
                 )
             self.graph = _init_pipeline_graph(root_node_name=candidate_roots[0])
 
-        component_definitions = get_component_definitions(pipeline_config=self.get_config())
-
-        # Check for duplicates before adding the definition
-        if name in component_definitions.keys():
+        # Check for duplicate names before adding the component
+        # Note that the very same component must be addable multiple times:
+        # E.g. for indexing pipelines it's common to add a retriever first and a document store afterwards.
+        # The document store is already being used by the retriever however.
+        # Thus the very same document store will be added twice, first as a subcomponent of the retriever and second as a first level node.
+        if name in self.components and self.components[name] != component:
             raise PipelineConfigError(f"A node named '{name}' is already in the pipeline. Choose another name.")
-        component_definitions[name] = component._component_config
+
+        component_definitions = get_component_definitions(pipeline_config=self.get_config())
+        component_definition = {"params": component.get_params(), "type": component.type}
+        component_definitions[name] = component_definition
 
```

Comment:
I'm not sure I get why we're "rebuilding" the `_component_config` here. In the original code, these two lines were replaced by ` component_definitions[name] = component._component_config`. Isn't that the same?

Reply:
You'r right, I originally got confused and thought the name was included in the component_definition, too. But that's not the case so `component._component_config` really is the same.